[RFC 0062] Content-addressed paths (#62)
NixOS/rfcs: rfcs/0062-content-addressed-paths.md

- We compute a cryptographic hash `$chash` of `$out`[^modulo-hashing]
- We move `$out` to `/nix/store/$chash-$name`
- We create an alias path from `$out` to `/nix/store/$chash-$name`

[^modulo-hashing]:

  We can possibly normalize all the self-references before
  computing the hash and rewrite them when moving the path to handle paths with
  self-references, but this isn't strictly required for a first iteration

### Building a normal derivation

The process for building a normal derivation is the following:

- We look into the drv for all the inputs paths of the build
- For each input path, we look whether the path is an alias. If so we replace it
  by its target
- We compute the `dynamic` output of the derivation from the patched version
- We then try to substitute and build the new derivation
- We create an alias path from the `static` output to the `dynamic` one

## Wrapping it up

# Drawbacks

[drawbacks]: #drawbacks

- Obviously, this makes the Nix model more complicated than what it is now. In
  particular, the caching model needs some modifications (see [caching]);

- We specify that only a sub-category of derivations can safely be marked as

>   `contentAddressed`, but there's no way to enforce these restricitions;

Reviewer: It's a bigger problem than it might look like, as it means that trivial updates can break the CA marking for reasons not worth mentioning in the upstream changelog.

Reviewer 2: Definitely :)

Maybe that could be clearly stated, but the original scope of this work was to be able to mark very specific derivations that were clearly guaranteed to be deterministic, in which case the problem was less important

Reviewer: Well, the question «why not just propagate CA» shows that writing more is a good idea.

I do think that stressing the limitations in a few key places is also a nice thing to do (people should be able to apply RFC as passed, not what was intended and not what was discussed, after all… we should not treat ourselves worse than we treat computers!)

- This will probably be a breaking-change for some tooling since the output path
  that's stored in the `.drv` files doesn't correspond to the actual on-disk
  path the output will be stored in (because it might just be an alias for the
  other path)

# Alternatives

[alternatives]: #alternatives

[RFC 0017][] is another proposal with the
same end-goal. The big difference between these two is in the scope they cover:
RFC 0017 is about fundamentally changing the base model of Nix, while this
proposal suggests to make only the minimal amount of changes to the current
model to allow the content-adressed model to live in parallel (which would open
the way to a fully content-adressed store as RFC0017, but in a much more
incremental way).

Eventually this RFC should be subsumed by RFC0017.

# Unresolved questions

> [unresolved]: #unresolved-questions

Reviewer: Should we have functionality that allows to build a CA package twice with _different_ apparent output paths, and optionally with different parallelism settings? The build of the package obviously fails if the CA unification doesn't lead to the same result.

Should we mandate that Hydra uses this functionality? Should it be on by default?

Reviewer 3: Per https://github.com/NixOS/rfcs/pull/62/files#r357243841 I think we can deal with non-deterministic derivation just fine.

Reviewer: Well, for binary cache transparency it is much better if you can build something locally, then regain connectivity and fetch stuff from a cache, then fetch stuff from a different cache, then build some more locally, etc.

Reviewer 3: So in my mind that's equally risky with and without content addressable derivations. The only difference is one lets you know if something goes wrong, and one doesn't.

Reviewer: No. Build nondeterminism doesn't introduce significant behaviour changes, so as long as the expectations are not broken (yeah, we install you into this output path and your dependencies into those paths, and that is not going to change), it will be mostly usable. There are a few CPU-dependent optimisations from time to time, they are annoying.

With CA things are actually moved around, so even though everything would still work when assembled together, the assembling part will be failing. It is Nix, not the code that is built by Nix, that would fail to do things because of nondeterminism.

Reviewer 3: I think trying to keep going despite non-determinism is a misfeature. You can always evict your own CA mappings (can keep the builds themselves for easy "rollback") and align with `cache.reflex-frp.org` and keep going.

Reviewer: I think if things _are_ marked CA, of course it is a good idea to catch failures. But what you propose will not catch much, because a typical derivation is only built once (ever) by Hydra, later Hydra will use the binary cache. Also my proposal includes feeding different «apparent» output paths to the same build with the same dependencies, which has a better chance of discovering compressed self-references.

Reviewer 3: I am OK with running `--check` and trying different termporary output paths. Catching non-determinism I don't think is important, because it's really *clashes* that we care about. However, catching self-references is important as we have to be able to move the thing.

Reviewer: Well, varying the output paths is something that doesn't follow from anything Nix does, so it has to be spelled explicitly.

## Caching

[caching]: #caching

The big unresolved question is about the caching of content-adressed paths.
As [Eelco's phd thesis][nixphd] states it, caching ca paths raises a number of
questions when building that path is non-deterministic (because two different
stores can have two different outputs for the same path, which might lead to
some dependencies being duplicated in the closure of a dependency).
There exist some solutions to this problem (including one presented in Eelco's
thesis), but for the sake of simplicity, this RFC simply forbids to mark a
derivation as ca if its build is not deterministic (although there's no real
way to check that so it's up to the author of the derivation to ensure that it

> is the case).

Reviewer 3: I don't think we can skip this. If we track all the evaluation steps, we have all the information to ensure a binary cache isn't given anything that clashes with ourself. Maybe the first prototype will discover these errrors lazily, but it should discover them

long-time dream of the community − a design for that was already taking a whole
chapter in [Eelco's PHD thesis][nixphd].

This was never done because it represents a quite big change in Nix's model,
with some non-totally-solved implications (regarding the trust model in
particular).
Even without going all the way down to a fully intensional model, we can
make specific paths content-adressed, which can give some important benefits of
the intensional store at a much lower price. In particular, setting some
critical derivations as content-adressed can lead to some substancial build
cutoffs.

# Detailed design

[design]: #detailed-design

In all that follows, we pretend that each derivation has only one output.
This doesn't change the reasoning but makes things easier to state.

The gist of the design is that:

- Some derivations can be marked as content-adressed (ca), in which case their
  output will be moved to a path `ca` determined only by its content after the
  build
- We introduce the notion of a `resolved derivation` which is a derivation that
  doesn't refer to any other derivation but only to concrete store paths.

>   To prevent ambiguities, we might speak of a `symbolic derivation` to

Reviewer 4: ```suggestion
  To prevent ambiguities, we will speak of a `symbolic derivation` to
```

There exist some solutions to this problem (including one presented in Eelco's
thesis), but for the sake of simplicity, this RFC simply forbids to mark a
derivation as ca if its build is not deterministic (although there's no real
way to check that so it's up to the author of the derivation to ensure that it
is the case).

## Client support

The bulk of the job here is done by the Nix daemon.

Depending on the details of the current Nix implementation, there might or
might not be a need for the client to also support it (which would require the
daemon and the client to be updated in synchronously)

## Old Nix versions and caching

What happens (and should happen) if a Nix not supporting the cas model queries
a cache with cas paths in it is not clear yet.

## Garbage collection

Another major open issue is garbage collection of the aliases table. It's not
clear when entries should be deleted. The paths in the domain are "fake" so we
can't use them for expiration. The paths in the codomain could be used (i.e. if
a path is GC'ed, we delete the alias entries that map to it) but it's not clear
whether that's desirable since you may want to bring back the path via

> substitution in the future.

Reviewer 3: I recommend we might just store the "generating" mappings from almost-resolved ca input paths (all deps resolved) to output paths, as this will require far less space. OTOH it makes garbage collection tricker as now all mappings in the build closure are needed to recover a maximum-unresolved input path to map.

Reviewer 3: OTOH we can just do that with tracing GC. We just read the table backwards, saying each derivation in the codomain references everything in the domain that maps to it, and then look those up in turn.

---
feature: Simple content-adressed store paths
start-date: 2019-08-14
author: Théophane Hufschmitt
co-authors: (find a buddy later to help our with the RFC)
shepherd-team: @layus, @edolstra and @Ericson2314

> shepherd-leader: (name to be appointed by RFC steering committee)

Reviewer 5: @regnat Can you add @edolstra as a leader here?

```suggestion
shepherd-leader: @edolstra
```

related-issues: (will contain links to implementation PRs)
---

# Summary

[summary]: #summary

Add some basic but simple support for content-adressed store paths to Nix.

We plan here to give the possibility to mark certain store paths as
content-adressed (ca), while keeping the other dependency-adressed as they are
now (modulo some mandatory drv rewriting before the build, see below)

By making this opt-in, we can impose arbitrary limitations to the paths that
are allowed to be ca to avoid some tricky issues that can arise with
content-adressability.

In particular, we restrict ourselves to paths that are:

> - without any non-textual self-reference (_i.e_ a self-reference hidden inside a zip file)

Reviewer 2: > I also don't understand the concept of "textual self-reference"

Indeed, I kinda gloss over it without explaining it anywhere. I'll add it somewhere. Thx

That way we don't have to worry about the fact that hash-rewriting is only an
approximation nor by the semantics of the distribution of non-deterministic
paths.

We also leave the option to lift these restrictions later.

This RFC already has a (somewhat working) POC at
<https://github.com/NixOS/nix/pull/3262>.

# Motivation

[motivation]: #motivation

Having a content-adressed store with Nix (aka the "Intensional store") is a
long-time dream of the community − a design for that was already taking a whole
chapter in [Eelco's PHD thesis][nixphd].

This was never done because it represents a quite big change in Nix's model,
with some non-totally-solved implications (regarding the trust model in
particular).
Even without going all the way down to a fully intensional model, we can
make specific paths content-adressed, which can give some important benefits of
the intensional store at a much lower price. In particular, setting some
critical derivations as content-adressed can lead to some substancial build
cutoffs.

# Detailed design

[design]: #detailed-design

The gist of the design is that:

- Derivations can be marked as content-adressed (ca), in which case each

>   one of their output will be moved to content-addressed `ca` store path.

Reviewer 2: Wow this is quite ambiguous indeed. The intent was to say that the output would be moved to a store path called `ca`, but as we don't reuse the name anywhere it's not useful

  content after the build
- We introduce the notion of a `resolved derivation` which is a derivation that

>   doesn't refer to any other derivation but only to concrete store paths.

Reviewer 3: > What kind of derivation are we talking about here, derivation as in the builtin in the expression language or .drv files?

Neither? but closer to the files. The concept (node in a merkle dag with certain data) exists ontologically prior to being a primop in some language or having some serialization format.

>   doesn't refer to any other derivation but only to concrete store paths.

Reviewer 3: > Also, what’s a concrete store path?

Good question. A content address store path, as exists today (fixed output, add-to-store, etc).

  To prevent ambiguities, we might speak of a `symbolic derivation` to

>   designate a derivation that's not necessarily resolved.

Reviewer 2: It is indeed the set of all derivations. But we don't use it anywhere so I'll drop it

  We also define a `resolving` function that given a symbolic derivation
  returns a new resolved derivation with the same semantics.
- When asked to build a derivation, Nix will first resolve it, build the
  resolved derivation and link back the symbolic one to the out path of the

>   resolved one.

Reviewer 2: Nah, `out` isn't intentional, thanks

## Nix-build process

### Output mappings

For each output `output` of a derivation `drv`, we define

> - its output id **DrvOutputId(drv, output)** as the tuple `(hash(drv), output, truster)`, where `truster` is a reserved field for future use and currently always set to `"world"`.

Reviewer 2: The reason for that was to prevent another schema change. That being said we might omit this in the RFC as it's indeed not relevant for the current design

  This id uniquely identifies the output.
  We textually represent this as `hash(drv)!output[@truster]`.
- its concrete path **PathOf(outputId)** as the path on which the output will be stored on disk.

> Unresolved: should we already include the `truster` field in `DrvOutputId`
> even if it's not used atm? What would be the cost of adding it later?

> In a input-addressed-only world, the concrete path for a derivation output was a pure function of this output's id that could be computed at eval-time. However this won't be the case anymore once we allow content-addressed derivations, so we now need to store the results the `PathOf` function in the Nix database as a new table:

Reviewer 4: ```suggestion
In a input-addressed-only world, the concrete path for a derivation output was a pure function of this output's id that could be computed at eval-time. However this won't be the case anymore once we allow CA derivations, so we now need to store the results of the `PathOf` function in the Nix database as a new table:
```

> In a dependency-addressed-only world, the concrete path for a derivation output was a pure function of this output's id that could be computed at eval-time. However this won't be the case anymore once we allow content-addressed derivations, so we now need to store the results the `PathOf` function in the Nix database as a new table:

Reviewer 6: Is there a difference between `dependency-addressed` and `content-addressed`? If no, we should stick with one term.

Reviewer 6: Ah, `dependency-addressed` is the term for the status of nix as-is? If you want to introduce this terminology here, you have to also define it somehow.

> In a dependency-addressed-only world, the concrete path for a derivation output was a pure function of this output's id that could be computed at eval-time. However this won't be the case anymore once we allow content-addressed derivations, so we now need to store the results the `PathOf` function in the Nix database as a new table:

Reviewer 6: ```suggestion
In a dependency-addressed-only world, `PathOf` was a pure function of `outputId` that could be computed at eval-time. However this won't be the case anymore once we allow content-addressed derivations, so we now need to store the results the `PathOf` function in the Nix database as a new table:
```

```sql
create table if not exists PathOf (
    drv integer not null,
    output text not null,
    truster integer not null,
    path integer not null,
)
```

> ### Building a normal derivation

Reviewer 6: where “normal” ≡ “symbolic”?

#### Resolved derivations

We define a **resolved derivation** as a derivation whose only references are either:

> - Placeholders for the its own outputs (from the `placeholder` builtin)

Reviewer 4: ```suggestion
- Placeholders for its own outputs (from the `placeholder` builtin)
```

> - References to the outputs of other (non content-addresed) resolved derivations

Reviewer 4: ```suggestion
- References to the outputs of other (non CA) resolved derivations
```

For a derivation `drv` whose input derivations have all been realised, we define its **associated resolved derivation** `resolved(drv)` as `drv` in which we replace every input derivation `inDrv` of `drv` by `pathOf(inDrv)` (and update the output hash accordingly).

> This doesn't have the property that for a derivation that doesn't depend on any CA derivation `resolved(drv) == drv`. I think that this is a rather big issue so we'll have to find a way to get this property back (but feel free to correct me if you think that it isn't a big deal)

`resolved` is (intentionally) not injective: If `drv` and `drv'` only differ because one depends on `dep` and the other on `dep'`, but `dep` and `dep'` are content-addressed and have the same output hash, then `resolved(drv)` and `resolved(drv')` will be equal.

#### Build process

When asked to build a derivation `drv`, we instead:

1. Compute `resolved(drv)`
2. Substitute and build `resolved(drv)` like a normal derivation.
   Possibly this is a no-op because it may be that `resolved(drv)` has already been built.
3. Add a new mapping `pathOf(drv!${output}) == ${output}(resolved(drv))` for each output `output` of `drv`

> ### Building a ca derivation

Reviewer 4: ```suggestion
### Building a CA derivation
```

> A **ca derivation** is a derivation with the `__contentAddressed` argument set

Reviewer 4: ```suggestion
A **CA derivation** is a derivation with the `__contentAddressed` argument set
```

   - We realise `resolved(contentAddressed'.drv)`. This gives us an output path `out(resolved(contentAddressed'.drv))`
   - We compute `ca(contentAddressed'.drv)` and notice that the path already exists (since it's the same as the one we built previously), so we discard the result.
   - We register in the db that `pathOf(contentAddressed.drv'!out) == ca(contentAddressed'.drv)` ( also equals to `ca(contentAddressed.drv)`)
3. We build `dependent'.drv`
   - We first compute `resolved(dependent'.drv)`.
     This gives us a new derivation identical to `dependent'.drv`, except that `contentAddressed'.drv!out` is replaced by `pathOf(contentAddressed'.drv!out) == ca(contentAddressed'.drv)`
   - We notice that `resolved(dependent'.drv) == resolved(dependent.drv)` (since `ca(contentAddressed'.drv) == ca(contentAddressed.drv)`), so we just return the already existing path
4. We build `transitivelyDependent'.drv`
   - We first compute `resolved(transitivelyDependent'.drv)`
   - Here again, we notice that `resolved(transitivelyDependent'.drv)` is the same as `resolved(transitivelyDependent.drv)`, so we don't build anything

# Drawbacks

[drawbacks]: #drawbacks

> - Obviously, this makes the Nix model more complicated than what it is now. In

Reviewer 4: ```suggestion
- Obviously, this makes the Nix model more complicated than it currently is. In
```

  particular, the caching model needs some modifications (see [caching]);

- We specify that only a sub-category of derivations can safely be marked as
  `contentAddressed`, but there's no way to enforce these restricitions;

- This will probably be a breaking-change for some tooling since the output path
  that's stored in the `.drv` files doesn't correspond to an actual on-disk
  path.

# Alternatives

[alternatives]: #alternatives

[RFC 0017][] is another proposal with the
same end-goal. The big difference between these two is in the scope they cover:
RFC 0017 is about fundamentally changing the base model of Nix, while this
proposal suggests to make only the minimal amount of changes to the current
model to allow the content-adressed model to live in parallel (which would open
the way to a fully content-adressed store as RFC0017, but in a much more
incremental way).

Eventually this RFC should be subsumed by RFC0017.

# Unresolved questions

[unresolved]: #unresolved-questions

## Caching

[caching]: #caching

The big unresolved question is about the caching of content-adressed paths.

> As [Eelco's phd thesis][nixphd] states it, caching ca paths raises a number of

Reviewer 4: ```suggestion
As [Eelco's phd thesis][nixphd] states, caching CA paths raises a number of
```

questions when building that path is non-deterministic (because two different
stores can have two different outputs for the same path, which might lead to
some dependencies being duplicated in the closure of a dependency).
There exist some solutions to this problem (including one presented in Eelco's
thesis), but for the sake of simplicity, this RFC simply forbids to mark a

> derivation as ca if its build is not deterministic (although there's no real

Reviewer 4: ```suggestion
derivation as CA if its build is not deterministic (although there's no real
```

way to check that so it's up to the author of the derivation to ensure that it
is the case).

## Client support

The bulk of the job here is done by the Nix daemon.

Depending on the details of the current Nix implementation, there might or

> might not be a need for the client to also support it (which would require the

Reviewer 4: ```suggestion
be a need for the client to also support it (which would require the
```

daemon and the client to be updated in synchronously)

## Old Nix versions and caching

> What happens (and should happen) if a Nix not supporting the cas model queries

Reviewer 4: *cas* hasn't been defined yet.

a cache with cas paths in it is not clear yet.

## Garbage collection

Another major open issue is garbage collection of the aliases table. It's not
clear when entries should be deleted. The paths in the domain are "fake" so we
can't use them for expiration. The paths in the codomain could be used (i.e. if
a path is GC'ed, we delete the alias entries that map to it) but it's not clear
whether that's desirable since you may want to bring back the path via
substitution in the future.

## Ensuring that no temporary output path leaks in the result

> One possible issue with the ca model is that the output paths get moved after being built, which breaks self-references. Hash rewriting solves this in most cases, but it is only heuristic and there is no way to truly ensure that we don't leak a self-reference (for example if a self-reference appears in a zipped file − like it's often the case for man pages or java jars, the hash-rewriting machinery won't detect it).

Reviewer 4: ```suggestion
One possible issue with the CA model is that the output paths get moved after being built, which breaks self-references. Hash rewriting solves this in most cases, but it is only a heuristic and there is no way to truly ensure that we don't leak a self-reference (for example if a self-reference appears in a zipped file − like is often the case for man pages or Java jars, the hash-rewriting machinery won't detect it).
```

> Having leaking self-references is annoying since

Reviewer 4: ```suggestion
Having leaking self-references is annoying since:
```

> - These self-references change each time the inputs of the derivation change, making ca useless (because the output will _always_ change when the input change)

Reviewer 4: ```suggestion
- These self-references change each time the inputs of the derivation change, making CA useless (because the output will _always_ change when the input change)
```

- More annoyingly, these references become dangling and can cause runtime failures

We however have a way to dectect these: If we have leaking self-references then the output will change if we artificially change its output path. This could be integrated in the `--check` option of `nix-store`.

# Future work

[future]: #future-work

This RFC tries as much as possible to provide a solid foundation for building

> ca paths with Nix, leaving as much room as possible for future extensions.

Reviewer 4: ```suggestion
CA paths with Nix, leaving as much room as possible for future extensions.
```

In particular:

- Add some path-rewriting to allow derivations with self-references to be built

>   as ca

Reviewer 4: ```suggestion
  as CA
```

- Consolidate the caching model to allow non-deterministic derivations to be

>   built as ca

Reviewer 4: ```suggestion
  built as CA
```

---
feature: Simple content-adressed store paths
start-date: 2019-08-14
author: Théophane Hufschmitt
co-authors: (find a buddy later to help our with the RFC)
shepherd-team: "@layus, @edolstra and @Ericson2314"
shepherd-leader: "@edolstra"
related-issues: (will contain links to implementation PRs)
---

# Summary

[summary]: #summary

> Add some basic but simple support for content-adressed store paths to Nix.

Reviewer 7: ```suggestion
Add some basic but simple support for content-addressed store paths to Nix.
```
is this correct? then change it everywhere

We plan here to give the possibility to mark certain store paths as
content-adressed (ca), while keeping the other input-adressed as they are

> now (modulo some mandatory drv rewriting before the build, see below)

Reviewer 7: ```suggestion
now (modulo some mandatory drv rewriting before the build, see below).
```

By making this opt-in, we can impose arbitrary limitations to the paths that
are allowed to be ca to avoid some tricky issues that can arise with
content-adressability.

In particular, we restrict ourselves to paths that only include textual
self-references (_e.g._ no self-reference hidden inside a zip file).

That way we don't have to worry about the fact that hash-rewriting is only an

> approximation

Reviewer 7: ```suggestion
approximation.
```

        else resolved(derivation, inputRealisations)

    if (derivationToBuild.isContentAddressed()):
        assignScratchOutputPaths(derivationToBuild)
        runBuildScript(derivationToBuild)
        return moveToCAPaths(derivationToBuild.outputs)
    else:
        runBuildScript(derivationToBuild)
        # If the derivation isn’t content-addressed, then it already knows its
        # own output paths
        return derivationToBuild.outputs
```

### Remote caching

#### Basic principles

A consequence of this change is that a store path is now just a meaningless
blob of data if it doesn't have its associated `realisation` metadata −
besides, Nix can't know the output path of a content-addressed derivation
before building it anymore, so it can't ask the remote store for it.

As a consequence, the remote cache protocols is extended to not simply
work on store paths, but rather at the realisation level:

- The store interface now specifies a new method

  ```python
  def queryRealisation(output : DrvOutput) -> Maybe Realisation
  ```

  If the store knows about the given derivation output, it will return the associated realisation, otherwise it will return `None`.

> - The substitution loop in Nix fist calls this method to ask the remote for the

Reviewer 8: ```suggestion
- The substitution loop in Nix first calls this method to ask the remote for the
```

[design]: #detailed-design

*In everything that follows, most algorithms and data-structures will be expressed as pseudo-python snippets*

When it comes to computing the output paths of a derivation, the current Nix
model, known as the “input-addressed” model (also sometimes referred to as the
“extensional” model) works (roughly) as follows:

1. A Nix language expression gets evaluated to a `derivation`
2. This `derivation` is a data-structure describing how to build a package. In particular it contains

>   1. A set of derivation outputs which will be used as input for the build

Reviewer 2: Yes, there’s a bit of nesting here that makes things a bit tricky to explain. It refers to the `inputDrvs` key (which is a set of derivation associated with a list of output names. Not sure how to word this better, any suggestion is wecome :) ).

>   1. A set of derivation outputs which will be used as input for the build

Reviewer 9: The problematic part is just that it wasn't clear to me at first glance what's the difference between 1. and 2. (what is the difference between a derivation output vs. store path). I suppose this might also be a bit ambiguous even when taking nixphd as the main reference...

> Figure 2.12 outlines this two-stage build process: Nix expressions are first translated to store derivations that live in the Nix store and that each describe a single build action with all variability removed. These store derivations can then be built, which results in derivation outputs that also live in the Nix store.

(nixphd, p. 39)

> The idea is that pre-built derivation outputs are made available in some central repository accessible to the client machines

(nixphd, p. 45)

are able to take into account the specificies of these new derivations.

Fully supporting content-addressed derivations requires some deep changes to the Nix model.
For the sake of readability, we’ll first present a simplistic model that support them in a very basic way, and then extend this model in several different ways to improve the support.

## Basic support

The input-addressed build process is roughly the following:

```python
def nix_build(expr : NixExpr) -> [StorePath] :
    resulting_derivation = eval(expr)
    build_derivation(
        resulting_derivation,
        resulting_derivation.all_outputs(),
    )
    return resulting_derivation.all_output_paths()

def build_derivation(derivation : Derivation, outputsToBuild: [str]) -> ():
    # Build all the inputs
    for (inputDrv, requiredOutputs) in derivation.inputDrvs:
        build_derivation(inputDrv, requiredOutputs)
    # Run the build script, now that all the inputs are here
    runBuildScript(derivation)
```

The main change required by the content-addressed model is that we can’t know
the output paths of a derivation before building it.

This means that the Derivations as they are produced by the evaluator can’t

> either know their output path, nor explicitely refer to their dependencies by

Reviewer 2: > So this is about unresolved derivations in contrast to resolved derivations which imo should be stated here

Yes, that’s why I precise “as they are produced by the evaluator”. I’ll emphasize this a bit more to make it clearer

> either know their output path, nor explicitely refer to their dependencies by

Reviewer 9: so, *evaluator* refers to the program which translates nix-lang code into unresolved nix-drvs. ok (as *evaluator* might be ambiguous with the *resolver* which translates unresolved drvs into resolved drvs).

In all that follows, we pretend that each derivation has only one output.
This doesn't change the reasoning but makes things easier to state.

The gist of the design is that:

- Some derivations can be marked as content-adressed (ca), in which case their
  output will be moved to a path `ca` determined only by its content after the
  build
- We introduce the notion of a `resolved derivation` which is a derivation that

>   doesn't refer to any other derivation but only to concrete store paths.

Reviewer 3: Unless we commit to breaking self cycles in the first version, we cannot actually do this because there is no hash of normal derivations with cyclic data.

> Derivations that don't transitively depend on any ca derivation are “equivalent” to their associated resolved derivation in that they refer to the same inputs and have the same output hash.

This quote in particular is in conflict with the above.

I think you wish to say that a `resolved` derivation depends on no content-addressed or fixed output derivations, but only concrete store paths and normal derivations.

- When asked to build a derivation, Nix will first resolve it, build the
  resolved derivation and link back the symbolic one to the out path of the
  resolved one.

## Nix-build process

### Output mappings

A major consequence of allowing content-addressed derivations is that the
actual output path of a derivation might not match its output hash anymore.

To express this, we introduce a new mapping `pathOf` that associates the hash
of every live derivation to its store path.
By extension, we also define `pathOf(drv) = pathOf(hash(drv))`

### Building a ca derivation

ca derivations are derivations with the `__contentAddressed` argument set to
`true`.

The process for building a content-adressed derivation is the following:

- We build it like a normal derivation (see below) to get an output path `$out`.
- We compute a cryptographic hash `$chash` of `$out`[^modulo-hashing]
- We move `$out` to `/nix/store/$chash-$name`
- We create a mapping from `$dhash` (the hash computed at eval-time) to
  `/nix/store/$chash-$name`

[^modulo-hashing]:

  We can possibly normalize all the self-references before
  computing the hash and rewrite them when moving the path to handle paths with

>   self-references, but this isn't strictly required for a first iteration

Reviewer 3: I would very much like to *not* do this the first drift.

Reviewer 10: Nix already supports this, behind a flag.

Reviewer 3: It does, but I'd still like to be a bit cautious. When the mounted store path is the content addressed thing, we open the door to all sorts of interesting ways to distribute and deploy nix stores. By adding this post-processing step we complicate those at the very least.

- its output id **DrvOutputId(drv, output)** as the tuple `(hash(drv), output, truster)`, where `truster` is a reserved field for future use and currently always set to `"world"`.
  This id uniquely identifies the output.
  We textually represent this as `hash(drv)!output[@truster]`.
- its concrete path **PathOf(outputId)** as the path on which the output will be stored on disk.

> Unresolved: should we already include the `truster` field in `DrvOutputId`
> even if it's not used atm? What would be the cost of adding it later?

In a dependency-addressed-only world, the concrete path for a derivation output was a pure function of this output's id that could be computed at eval-time. However this won't be the case anymore once we allow content-addressed derivations, so we now need to store the results the `PathOf` function in the Nix database as a new table:

```sql
create table if not exists PathOf (
    drv integer not null,
    output text not null,
    truster integer not null,
    path integer not null,
)
```

### Building a normal derivation

#### Resolved derivations

We define a **resolved derivation** as a derivation whose only references are either:

> - Self references

Reviewer 3: Strictly speaking its the outputs, not the DRVs themselves, that have the self references. DRV self-references are handled with the `placeholder` thing which is different.

Reviewer 6: Yeah, the definition of a self-reference is missing here.

- References to the outputs of other (non content-addresed) resolved derivations

> - Existing store paths

Reviewer 3: ```suggestion
Existing store paths (including *built* content-addresed resolved derivations' output paths)
```

)
```

### Resolved derivations

As it is already internally the case in Nix, we define a **basic derivation** as a derivation that doesn't depend on any derivation output (except its own). Said otherwise, a basic derivation is a derivation whose only inputs are either

- Placeholders for its own outputs (from the `placeholder` builtin)
- Existing store paths

For a derivation `drv` whose input derivations have all been realised, we define its **associated resolved derivation** `resolved(drv)` as `drv` in which

> we replace every input derivation `inDrv` of `drv` by `Realisation(inDrv).path`.

Reviewer 9: I suppose the newline shouldn't be inserted.

Reviewer 2: I’ve never been able to choose beween a strict 80 chars wrapping and wrapping at the end of each sentence when writing markdown code. That one is a weird mix that results in following neither of these options 😑

`resolved` is (intentionally) not injective: If `drv` and `drv'` only differ because one depends on `dep` and the other on `dep'`, but `dep` and `dep'` are content-addressed and have the same output hash, then `resolved(drv)` and `resolved(drv')` will be equal.

### content-addressed build process

We now need to update the build process as:

```python
def build_derivation(derivation : Derivation, outputsToBuild: [str]) -> Map[DrvOutput, Realisation]:
    inputRealisations : Map[DrvOutput, Realisation] = {}
    # Build all the inputs, and store the newly built realisations
    for (inputDrv, requiredOutputs) in derivation.inputDrvs:

>         inputRealisations += build_derivation(inputDrv, requiredOutputs)

Reviewer 9: this line is a bit confusing... where does the key for the new entry in `inputRealisations`, which is a associative array afaik, come from?

Reviewer 2: `build_derivation` also returns a map. Would something like

```python
inputRealisations.extend(build_derivation(inputDrv, requiredOutputs))
```

be clearer?

Reviewer 9: yes, thanks.

It’s however possible to freely mix content- and input-addressed derivations in the same Nix store, and even in the same closure:

The algorithm for building content-addressed derivations extends the algorithm for building input-addressed derivations in two ways:
1. Before running the build script, it resolves the derivation
2. When running the build script, it uses some temporary outputs, and moves them to their final location afterwards.

Only the second part assumes that the derivation is content-addressed, and we can use two-different code-paths for the build-step:

```python
def build_derivation(derivation : Derivation, outputsToBuild: [str]) -> Map[DrvOutput, Realisation]:
    # Build the dependencies and resolve the derivation like before
    derivationToBuild = ...

    if (derivationToBuild.isContentAddressed()):
        assignScratchOutputPaths(derivationToBuild)
        runBuildScript(derivationToBuild)
        return moveToCAPaths(derivationToBuild.outputs)
    else:
        runBuildScript(derivationToBuild)
        # If the derivation isn’t content-addressed, then it already knows its
        # own output paths

>         return derivationToBuild.outputs()

Reviewer 9: ```suggestion
        return derivationToBuild.outputs
```

This means that it also has to change the output paths (as these depend on the inputs of the derivation).

That’s something that we don’t want for the derivations that are already valid today, so we must bypass the resolving step for these derivations (which is okay as these derivations don’t need to be resolved).

```python
def build_derivation(derivation : Derivation, outputsToBuild: [str]) -> Map[DrvOutput, Realisation]:
    inputRealisations : Map[DrvOutput, Realisation] = {}
    # Build all the inputs, and store the newly built realisations
    for (inputDrv, requiredOutputs) in derivation.inputDrvs:
        inputRealisations += build_derivation(inputDrv, requiredOutputs)

    derivationToBuild =
        derivation if derivation.isStrictlyInputAddressed()
        else resolved(derivation, inputRealisations)

    if (derivationToBuild.isContentAddressed()):
        assignScratchOutputPaths(derivationToBuild)
        runBuildScript(derivationToBuild)
        return moveToCAPaths(derivationToBuild.outputs)
    else:
        runBuildScript(derivationToBuild)
        # If the derivation isn’t content-addressed, then it already knows its
        # own output paths

>         return derivationToBuild.outputs()

Reviewer 9: ```suggestion
        return derivationToBuild.outputs
```

work on store paths, but rather at the realisation level:

- The store interface now specifies a new method
  ```python
  def queryRealisation(output : DrvOutput) -> Maybe Realisation
  ```

  If the store knows about the given derivation output, it will return the associated realisation, otherwise it will return `None`.
- The substitution loop in Nix fist calls this method to ask the remote for the
  realisation of the current derivation output.
  If this first call succeeds, then it fetches the corresponding output path
  like before. Then, it registers the realisation in the database:

  ```python
  def substitute_realisation(substituter : Store, wantedOutput : DrvOutput) -> Maybe Realisation:
      maybeRealisation = substituter.queryRealisation(wantedOutput)

>       if not maybeRealisation:

Reviewer 9: ```suggestion
      if maybeRealisation is None:
```

- Realisations now contain a `dependencies` field, which is a map from `DrvOutput` to `StorePath`:

    ```python
    class Realisation:
        id : DrvOutput
        outputPath : StorePath
        dependencies : Map[DrvOutput, StorePath]
    ```
- We add the constraint that realisations should form a closure in a store, meaning that if a store has the realisation for `foo!out` with a dependency on `bar!out->/nix/store/bar`, then the store must also have a realisation for `bar!out` whose output path is `/nix/store/bar`
- The realisation loop now keep tracks of these realisations to enforce this closure invariant:
  ```python
  # Returns true (and warns) iff we already have a realisation for the given
  # derivation output, and that realisation has a different output path
  # than the expected one.
  def is_incompatible(drvOutput, expectedStorePath):
      maybeLocalRealisation = localStore.queryRealisation(drvOutput)
      if (maybeLocalRealisation and maybeLocalRealisation.outputPath != expectedStorePath):
          warn(f"The substituter {substituter} has an incompatible realisation for {dependentDrvOutput}")

>           return true
>       return false

Reviewer 9: ```suggestion
          return True
      return False
```

  def substitute_realisation(substituter : Store, wantedOutput : DrvOutput) -> Maybe Realisation:
      maybeRealisation = substituter.queryRealisation(wantedOutput)

>       if not maybeRealisation:

Reviewer 9: ```suggestion
      if maybeRealisation is None:
```

Eelco's thesis), but there are much more complex, so it's probably not worth
investing in them until we're sure that they are needed.

# Future work

[future]: #future-work

This RFC tries as much as possible to provide a solid foundation for building

> ca paths with Nix, leaving as much room as possible for future extensions.

Reviewer 9: ```suggestion
CA paths with Nix, leaving as much room as possible for future extensions.
```
